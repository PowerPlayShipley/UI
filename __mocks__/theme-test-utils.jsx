// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
