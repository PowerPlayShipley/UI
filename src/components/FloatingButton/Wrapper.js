import styled from "styled-components";
import { tint, parseToRgb, rgbToColorString } from "polished";

const Wrapper = styled.button`
  box-sizing: border-box;

  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  
  border: none;
  border-radius: 50%;
  box-shadow: 0 3px 5px -1px rgb(0 0 0 / 20%), 
    0 6px 10px 0 rgb(0 0 0 / 14%), 
    0 1px 18px 0 rgb(0 0 0 / 12%);
  
  padding: 0;
  width: ${({ theme }) => theme.main.sizes.scale['size-7']};
  height: ${({ theme }) => theme.main.sizes.scale['size-7']};
  
  cursor: pointer;
  user-select: none;
  
  overflow: hidden;

  color: ${({ color, theme }) => color ? color : theme.main.colors.system.white};
  background-color: ${({ background, theme }) => background ? background : theme.main.colors.defaults.greenBright};

  & {
    background-position: center;
    transition: background 0.4s;
  }
  
  &:hover &:active {
    box-shadow: 0 5px 5px -3px rgb(0 0 0 / 20%), 
      0 8px 10px 1px rgb(0 0 0 / 14%), 
      0 3px 14px 2px rgb(0 0 0 / 12%);
  }

  &:disabled, &.disabled, &[aria-disabled=true] {
    color: ${({ theme }) => rgbToColorString({...parseToRgb(theme.main.colors.system.white), alpha: 0.9})} !important;
    background-color: ${({ background, theme }) => tint(0.3, background ? background : theme.main.colors.defaults.greenBright)} !important;
    
    cursor: default !important;
  }
  
  // TODO: Set these to work with light and dark colors
  &:hover {
    --hover-background: ${({ background, theme }) => tint(0.2, background ? background : theme.main.colors.defaults.greenBright)};
  }
  
  &:hover {
    cursor: pointer;
    background: var(--hover-background) radial-gradient(circle, transparent 1%, var(--hover-background) 1%) center/15000%;
  }

  &.ripple:active {
    --active-background: ${({ background, theme }) => tint(0.4, background ? background : theme.main.colors.defaults.greenBright)};
  }
  
  &.ripple:active {
    background-color: var(--active-background);
    background-size: 100%;
    transition: background 0s;
  }
`

export default Wrapper
