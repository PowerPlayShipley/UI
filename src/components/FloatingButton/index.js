import React from 'react'
import PropTypes from 'prop-types'
import clsx from "clsx";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons"
import Wrapper from "./Wrapper";

const FloatingButton = ({ className, ripple = false, icon = faPlus, ...rest}) => {
  return (
    <Wrapper className={clsx(ripple && 'ripple', className)} {...rest}>
      <FontAwesomeIcon icon={icon} />
    </Wrapper>
  )
}

FloatingButton.propTypes = {
  background: PropTypes.string,
  className: PropTypes.string,
  color: PropTypes.string,
  disabled: PropTypes.bool,
  ripple: PropTypes.bool,
  onClick: PropTypes.func,
  icon: PropTypes.any.isRequired
}

export default FloatingButton
