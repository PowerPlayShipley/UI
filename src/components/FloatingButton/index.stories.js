import { faPlus } from "@fortawesome/free-solid-svg-icons"

import FloatingButton from "./index";

export default {
  title: 'Button/Floating',
  component: FloatingButton
}

const Template = (args) => (<FloatingButton {...args} />);

export const Default = Template.bind({});
Default.args = {
  icon: faPlus,
  ripple: true
}

export const Disabled = Template.bind({});
Disabled.args = {
  icon: faPlus,
  disabled: true,
  ripple: true
}

export const NoneRipple = Template.bind({});
NoneRipple.args = {
  icon: faPlus,
  ripple: false
}
