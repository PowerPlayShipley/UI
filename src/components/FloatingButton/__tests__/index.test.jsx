/**
 * @jest-environment jsdom
 */

import React from 'react'
import { faPlus } from "@fortawesome/free-solid-svg-icons"

import { renderWithTheme, fireEvent } from '../../../../__mocks__/theme-test-utils'

import FloatingButton from "../index";

describe('Display Button', () => {
  it('should render button', () => {
    const { container } = renderWithTheme(<FloatingButton icon={faPlus} />)

    const svg = container.querySelector('svg')
    expect(svg).toHaveClass('fa-plus')
  })

  it('should create valid snapshot', function () {
    const { container } = renderWithTheme(<FloatingButton icon={faPlus} />)
    expect(container.firstChild).toMatchSnapshot()
  });

  it('should fire button', function () {
    const handleClick = jest.fn()

    const { container } = renderWithTheme(<FloatingButton icon={faPlus} onClick={handleClick} />)

    fireEvent.click(container.querySelector('button'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  });
})
