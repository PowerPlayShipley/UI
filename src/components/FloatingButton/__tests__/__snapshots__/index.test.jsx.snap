// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Button should create valid snapshot 1`] = `
.c1 {
  box-sizing: border-box;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: none;
  border-radius: 50%;
  box-shadow: 0 3px 5px -1px rgb(0 0 0 / 20%), 0 6px 10px 0 rgb(0 0 0 / 14%), 0 1px 18px 0 rgb(0 0 0 / 12%);
  padding: 0;
  width: 48px;
  height: 48px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  color: #ffffff;
  background-color: #1a7f37;
}

.c1 {
  background-position: center;
  -webkit-transition: background 0.4s;
  transition: background 0.4s;
}

.c1:hover .c0:active {
  box-shadow: 0 5px 5px -3px rgb(0 0 0 / 20%), 0 8px 10px 1px rgb(0 0 0 / 14%), 0 3px 14px 2px rgb(0 0 0 / 12%);
}

.c1:disabled,
.c1.disabled,
.c1[aria-disabled=true] {
  color: rgba(255,255,255,0.9) !important;
  background-color: #5ea573 !important;
  cursor: default !important;
}

.c1:hover {
  --hover-background: #47985f;
}

.c1:hover {
  cursor: pointer;
  background: var(--hover-background) radial-gradient(circle,transparent 1%,var(--hover-background) 1%) center/15000%;
}

.c1.ripple:active {
  --active-background: #75b287;
}

.c1.ripple:active {
  background-color: var(--active-background);
  background-size: 100%;
  -webkit-transition: background 0s;
  transition: background 0s;
}

<button
  class="c0 c1"
>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-plus fa-w-14 "
    data-icon="plus"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 448 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      fill="currentColor"
    />
  </svg>
</button>
`;
